mitiq/benchmarks/mirror_circuits.py

```diff
@@ -0,0 +1,160 @@
+    are to be applied given a two qubit gate density
+    and the connectivity graph that must be satisfied.
+
+    Args:
+        two_qubit_gate_prob: Probability of an edge being chosen
+            from the set of candidate edges.
+            connectivity_graph_: The connectivity graph for the backend
+            on which the circuit will be run.
+    """
+    connectivity_graph = connectivity_graph.copy()
+    candidate_edges = nx.Graph()
+
+    final_edges = nx.Graph()
+    final_edges.add_nodes_from(connectivity_graph)
+
+    while connectivity_graph.edges:
+        num = random.randint(connectivity_graph.size())
+        edges = list(connectivity_graph.edges)
+        curr_edge = edges[num]
+        candidate_edges.add_edge(*curr_edge)
+        connectivity_graph.remove_nodes_from(curr_edge)
+
+    for edge in candidate_edges.edges:
+        if random.random() < two_qubit_gate_prob:
+            final_edges.add_edge(*edge)
+    return final_edges
+
+
+def random_cliffords(connectivity_graph: nx.Graph) -> cirq.Circuit:
+    """Returns a circuit with a two-qubit Clifford gate applied
+    to each edge in edges, and a random single-qubit
+    Clifford gate applied to every other qubit.
+
+    Args:
+        edges: A graph with the edges for which the
+            two-qubit Clifford gate is to be applied.
+    """
+    gates = [
+        cirq.CNOT(cirq.LineQubit(a), cirq.LineQubit(b))
+        for a, b in list(connectivity_graph.edges)
+    ]
+    qubits = nx.Graph()
+    qubits.add_nodes_from(nx.isolates(connectivity_graph))
+    gates.append(random_single_cliffords(qubits))
+    return cirq.Circuit(gates)
+
+
+def random_single_cliffords(connectivity_graph: nx.Graph) -> cirq.Circuit:
+    """Returns a circuit with a random single-qubit Clifford gate
+    applied on each given qubit.
+
+    Args:
+        qubits: A graph with each node representing a qubit for
+            which a random single-qubit Clifford gate is to be applied.
+    """
+    gates: List[cirq.Operation] = []
+    for qubit in connectivity_graph.nodes:
+        num = random.randint(len(cliffords))
+        for clifford_gate in cliffords[num]:
+            gates.append(clifford_gate(cirq.LineQubit(qubit)))
+    return cirq.Circuit(gates)
+
+
+def generate_mirror_circuit(
+    nlayers: int,
+    two_qubit_gate_prob: float,
+    connectivity_graph: nx.Graph,
+    seed: Optional[int] = None,
+) -> cirq.Circuit:
+    """Returns a randomized mirror circuit.
+
+    Args:
+        nlayers: The number of random Clifford layers to be generated.
+            two_qubit_gate_prob: Probability of a two-qubit gate being applied.
+            connectivity_graph: The connectivity graph of the backend
+            on which the mirror circuit will be run. This is used
+            to make sure 2-qubit gates are only run on connected qubits.
+    """
+    if not 0 <= two_qubit_gate_prob <= 1:
+        raise ValueError("two_qubit_gate_prob must be between 0 and 1")
+
+    random.seed(seed)
```

Comment:
Use `random_state = np.random.RandomState(seed)` instead of a global seed. See [example here](https://github.com/unitaryfund/mitiq/blob/52afc73b4035ec4b907fcceedd3997396e9417a7/mitiq/zne/scaling/folding.py#L416-L417). This [short writeup](https://towardsdatascience.com/stop-using-numpy-random-seed-581a9972805f) explains why.

The functions where you use the random state, e.g. `random_paulis` and `edge_grab`, should then input this random state (or the seed) and use it.